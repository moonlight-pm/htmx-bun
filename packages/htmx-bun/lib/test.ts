import { afterAll } from "bun:test";
import { mkdtempSync, rmdirSync } from "node:fs";
import { tmpdir } from "node:os";

export function makeTemporaryDirectory() {
    const dir = mkdtempSync(`${tmpdir()}/htmx-bun-`);
    afterAll(() => {
        rmdirSync(dir, { recursive: true });
    });
    return dir;
}
