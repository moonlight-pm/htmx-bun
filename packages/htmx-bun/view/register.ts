import { Glob } from "bun";
import { resolve } from "path";
import { error, info, warn } from "~/lib/log";
import { Template, TemplateModule } from "./template";

export class TemplateRegister {
    templates: Record<string, Template> = {};

    constructor(
        // Base path for the view tree
        public base: string,
    ) {}

    async initialize() {
        for await (const path of new Glob("**/*.part").scan(this.base)) {
            const tag = tagFromPath(path);
            if (this.templates[tag]) {
                warn(
                    "view",
                    `Duplicate tag '${tag}' defined in '${this.base}/${path}'`,
                );
                warn(
                    "view",
                    `Using '${this.base}/${this.templates[tag].path}'`,
                );
                continue;
            }
            info("register", `registering partial at '${path}'`);
            await this.reload(path);
        }
    }

    async reload(path: string) {
        const tag = tagFromPath(path);
        const absoluteFilePath = resolve(`${this.base}/${path}`);
        if (this.templates[tag]) {
            delete require.cache[absoluteFilePath];
        }
        try {
            const module = (await import(absoluteFilePath)) as TemplateModule;
            this.templates[tag] = new Template(this, tag, path, module);
        } catch (e) {
            error("view", `Failed to load '${path}'`);
            // @ts-ignore
            const cause = e.cause?.toString();
            if (cause) {
                console.log(cause);
            }
            return;
        }
    }

    get(tag: string): Template | undefined {
        return this.templates[tag];
    }
}

function tagFromPath(path: string) {
    let tag = path
        .split("/")
        .filter((it) => !/^\(.*\)$/.test(it))
        .join("-")
        .replace(/\.part/g, "");
    if (tag === "index") {
        tag = "root";
    }
    if (tag.endsWith("-index")) {
        tag = tag.replace(/-index$/, "");
    }
    return tag;
}
