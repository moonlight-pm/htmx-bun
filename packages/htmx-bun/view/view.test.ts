import { expect, test } from "bun:test";
import { TemplateRegister } from "./register";

const register = new TemplateRegister("./view/__fixtures__");
await register.initialize();

test("render", async () => {
    const view = register.get("todo-item").present();
    const html = await view.render({ id: 1, name: "Luke" });
    expect(html).toMatchSnapshot();
});

test("composition", async () => {
    const view = register.get("container").present();
    const html = await view.render();
    expect(html).toMatchSnapshot();
});

test("interpolated composition", async () => {
    const view = register.get("root").present();
    const html = await view.render();
    expect(html).toMatchSnapshot();
});

test("local environment", async () => {
    const view = register.get("environ").present();
    const html = await view.render();
    expect(html).toMatchSnapshot();
});

test("attributes passed to local environment", async () => {
    const view = register.get("passattr").present();
    const html = await view.render();
    expect(html).toMatchSnapshot();
});
