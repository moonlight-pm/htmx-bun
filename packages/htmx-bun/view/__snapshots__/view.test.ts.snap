// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`render 1`] = `
"<li id="1">
    Luke
</li>
"
`;

exports[`composition 1`] = `
"<div>
    <div>
        Widget
    </div>
</div>
"
`;

exports[`interpolated composition 1`] = `
"<html>
    <body>
        <ul>
            <li id="1">
                Love
            </li>
            <li id="2">
                Joy
            </li>
            <li id="3">
                Peace
            </li>
            <li id="4">
                Patience
            </li>
            <li id="5">
                Kindness
            </li>
            <li id="6">
                Goodness
            </li>
            <li id="7">
                Faithfulness
            </li>
            <li id="8">
                Gentleness
            </li>
            <li id="9">
                Self-control
            </li>
        </ul>
    </body>
</html>
"
`;

exports[`local environment 1`] = `
"<div>
    He first loved us.
</div>
"
`;

exports[`attributes passed to local environment 1`] = `
"<div>
    And He is before all things, and in Him all things hold together.
</div>
<div>
    true
</div>
"
`;
